# 「Next.js 13 app directory で記事投稿サイトを作ってみよう」を読む

[Next\.js 13 app directory で記事投稿サイトを作ってみよう](https://zenn.dev/azukiazusa/articles/next-js-app-dir-tutorial)


## 起動

```
info  - VS Code settings.json has been created for Next.js' automatic app types, this file can be added to .gitignore if desired
```

## `app/page.tsx` 編集

> ファイルを編集した後 http://localhost:3000/ にアクセスすると、以下のように表示が切り替わっていることが確認できます。

`layout.tsx` を残した状態だとスタイルが適用される。

## `app/articles/[slug]/page.tsx` 作成

`app/articles/[slug]/page.tsx` を作る。

> 将来的に params の型は TypeScript plugin により自動で生成されるようになる予定です。

気になる👀

## `app/layout.tsx` 編集

> Next.js は `<html>` や `<body>` タグを自動的に生成しないので、`app/layout.tsx` ファイルで必ず定義する必要があります。

へー。


ここで怒られる

```ts
export const metadata = {
           //^^^^^^^^ "metadata" is not a valid Next.js entry export value.
  title: "Create Next App",
  description: "Generated by create next app",
  themeColor: "#ffffff",
};
```

[Getting a TypeScript error: "metadata" is not a valid next\.js entry export value : r/nextjs](https://www.reddit.com/r/nextjs/comments/11xqy91/getting_a_typescript_error_metadata_is_not_a/)

気になるがのエラーでページの表示に影響していないので一旦放置。
"Next.js entry export value"ってTSのエラーとして検知できるの？

## Chakra UI 導入

```
npm i @chakra-ui/react @emotion/react@^11 @emotion/styled@^11 framer-motion@^6
```

## app/lauout.tsx に ChakraProvider を追加

```
./node_modules/@chakra-ui/accordion/dist/chunk-JDQBKIKM.mjs

You're importing a component that needs useState. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.

    ,-[8:1]
  8 | import { useControllableState } from "@chakra-ui/react-use-controllable-state";
  9 | import { mergeRefs } from "@chakra-ui/react-use-merge-refs";
 10 | import { callAllHandlers, warn } from "@chakra-ui/shared-utils";
 11 | import { useCallback, useEffect, useId, useRef, useState } from "react";
    :                                                 ^^^^^^^^
 12 | function useAccordion(props) {
 13 |   const {
 14 |     onChange,
    `----

Maybe one of these should be marked as a client entry with "use client":
  node_modules\@chakra-ui\accordion\dist\chunk-JDQBKIKM.mjs
  node_modules\@chakra-ui\accordion\dist\index.mjs
  node_modules\@chakra-ui\react\dist\index.mjs
  app\layout.tsx
```

Server Component, `onClick`　すら使えないのか。テンプレートエンジンみたいだ。

## app/Provider.tsx 作成

`components` ディレクトリとかではなくて `page.tsx` と同階層に置いてしまう。

## ビルド

1回ビルドしてみる。

```
Route (app)                                Size     First Load JS
┌ ○ /                                      0 B                0 B
└ λ /articles/[slug]                       139 B          66.6 kB
+ First Load JS shared by all              66.4 kB
  ├ chunks/17-35779e5c4405374c.js          64.5 kB
└ λ /api/articles/[slug]/comments          0 B            81.1 kB
+ First Load JS shared by all              81.1 kB
  ├ chunks/main-07078917cb8e3d57.js        79.2 kB
  ├ chunks/pages/_app-5841ab2cb3aa228d.js  192 B
  └ chunks/webpack-c58b595510065273.js     1.75 kB

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
```

## fetch

> app/pages.tsx 内で http://localhost:3000/api/articles から記事の一覧を取得します。

`page.tsx` っぽい。

`Home` を async 関数（コンポーネント）として再定義する。

> デフォルトでは `fetch` を使用すると自動的にデータを取得した後にキャッシュされます。これは `fetch` のオプションはデフォルトで `{ cache: "force-cache" }` が設定されていることを意味します。`force-cache` オプションは `getStaticProp` と近い働きです。

ほぼ永続？　再検証タイミングは？

TODO: Next.js の `fetch` のドキュメント見る

`fetch` は型的には Node 標準のものと同じ。

## app/loading.tsx 追加

> この挙動は Suspense における `fallback` と同じです。

`page.tsx` の `resolve` が済むまで、 `layout.tsx` の `children` の部分丸ごとこれになる。

## app/error.tsx 追加

> また `error.tsx` は必ず Client Component として扱われます。

わざと `"use client";` を欠くと以下のエラー。ブラウザに何も出ない。

```
Error: Functions cannot be passed directly to Client Components because they're not serializable.
  <... parallelRouterKey=... segmentPath=... error={function} errorStyles=... loading=... loadingStyles=... hasLoading=... template=... templateStyles=... notFound=... notFoundStyles=... childProp=... rootLayoutIncluded=...>
                                                   ^^^^^^^^^^
    at resolveModelToJSON (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1766:39)
    at Object.toJSON (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1132:40)
    at stringify (<anonymous>)
    at processModelChunk (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:172:36)
    at retryTask (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1879:50)
    at performWork (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1917:33)
    at eval (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1308:40)    
    at scheduleWork (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:52:25)
    at pingTask (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1307:29)
    at ping (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1320:40)    
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Error: Functions cannot be passed directly to Client Components because they're not serializable.
  <... parallelRouterKey=... segmentPath=... error={function} errorStyles=... loading=... loadingStyles=... hasLoading=... template=... templateStyles=... notFound=... notFoundStyles=... childProp=... rootLayoutIncluded=...>
                                                   ^^^^^^^^^^
    at resolveModelToJSON (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1766:39)
    at Object.toJSON (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1132:40)
    at stringify (<anonymous>)
    at processModelChunk (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:172:36)
    at retryTask (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1879:50)
    at performWork (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1917:33)
    at eval (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1308:40)    
    at scheduleWork (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:52:25)
    at pingTask (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1307:29)
    at ping (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/server.browser.js:1320:40)    
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  digest: '4084611050'
}
```

## ArticleList コンポーネント作成

`app/components/ArticleCard.tsx` とあるがおそらく `app/ArticleCard.tsx`

`app/components/ArticleList.tsx` も `app/ArticleList.tsx`

うーん

```
Unhandled Runtime Error
Error: Hydration failed because the initial UI does not match what was rendered on the server.

See more info here: https://nextjs.org/docs/messages/react-hydration-error
```

↑が2回と、↓が1回。

```
Error: There was an error while hydrating this Suspense boundary. Switched to client rendering.
```

これが原因だろうか。

> Common causes with css-in-js libraries:
>
> When using Styled Components / Emotion

コンソールにこういうのも出ている：
```
Warning: Expected server HTML to contain a matching <h1> in <div>.
```

## コメントの取得

> `revalidate` は `fetch` のオプションとして渡すだけでなく、ルート単位の設定もできます。ルート単位で設定するには `page` または `layout` ファイルで以下のように記述します。
>
> ```ts
> export const revalidate = 60
> ```

ISRっぽい。

`app/pages/articles/[slug].tsx` とあるが、 `app/articles/[slug]/page.tsx`。
`app/pages/articles/not-found.tsx` とあるが、`app/articles/[slug]/not-found.tsx`

`fallback` とは別に404画面を出せる。
```ts
  if (res.status === 404) {
    // notFound 関数を呼び出すと not-fount.tsx を表示する
    notFound();
  }

  if (!res.ok) {
    throw new Error("Failed to fetch article");
  }
```

この時点で `/articles/hoge.tsx` を開くと、`Article` の名前（型名とページコンポーネント）が重複してエラーになった。

`app/pages/articles/[slug].tsx` とあるが、 `app/articles/[slug]/page.tsx`。

```diff
-  return data.comments as Comment[];
+  return data as Comment[];
```

Resolveされる前に `Comments` がレンダーされてしまう。

```
TypeError: Cannot read properties of undefined (reading 'map')
    at Comments (webpack-internal:///(sc_server)/./app/articles/[slug]/page.tsx:93:28)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
TypeError: Cannot read properties of undefined (reading 'map')
    at Comments (webpack-internal:///(sc_server)/./app/articles/[slug]/page.tsx:93:28)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  digest: '1623454996'
}
```

解決しないので保留。

落ちる……

```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11404:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async getArticle (C:\path\to\nextjs-app-dir-example\.next\server\app\articles\[slug]\page.js:427:17)
    at async ArticleDetail (C:\path\to\nextjs-app-dir-example\.next\server\app\articles\[slug]\page.js:455:21) {
  cause: Error: connect ECONNREFUSED ::1:3000
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    errno: -4078,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '::1',
    port: 3000
  }
}
[Error: An error occurred in the Server Components render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included on this error instance which may provide additional details about the nature of the error.] {
  digest: '2973016215'
}
node:internal/deps/undici/undici:11404
    Error.captureStackTrace(err, this);
          ^

TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11404:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async getComments (C:\path\to\nextjs-app-dir-example\.next\server\app\articles\[slug]\page.js:443:17) {
  cause: Error: connect ECONNREFUSED ::1:3000
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    errno: -4078,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '::1',
    port: 3000
  }
}

Node.js v18.14.1
error Command failed with exit code 1.
```

`package.json` を `complete` ブランチに揃えてみる→だめ
https://github.com/azukiazusa1/nextjs-app-dir-example/blob/complete/package.json

`complete`ブランチを動かしてみる→OK

ならNodeやパッケージのバージョンではなくてコードの問題だ。

---

あほだった…………。タイポだった。＞`commentPromise`と`commetsPromise`が混在。
```tsx
{/* @ts-expect-error 現状は jsx が Promise を返すと TypeScript が型エラーを報告するが、将来的には解決される */}
<Comments commentPromise={commentsPromise} />
```

```tsx
export default async function Comments({
    commentsPromise,
  }: {
    commentsPromise: Promise<Comment[]>;
  }) {
```
エラー抑制していたせいで未定義プロパティを使っていることに気付かず……。
というか未定義プロパティがあること自体はJS的には実行時エラーにならないのか。
ビルドも通ってしまうので気付けなかった。

## metadata

> ルーティングごとに `<head>` タグを設定するには、`page.tsx` または `layout.tsx` ファイル内で `metadata` オブジェクトまたは `generateMetadata` 関数を名前付きエクスポートします。

`generateMetadata` の型(`GenerateMetadata`)が欲しいが無い？`Metadata` 型はある。

> `ArticleDetail` コンポーネントと同じリクエストを送信しているので一見非効率なように思えますが、Next.js により `fetch` を利用したリクエストは自動で重複排除されるのでパフォーマンスには影響しません。

すごすぎる。

> "metadata" is not a valid Next.js entry export value.ts(71002)

https://github.com/vercel/next.js/issues/46431